File: thinking-orbs-mega/ports/ios/ThinkingOrbs/Sources/ThinkingOrbs/Modes/Lattice.swift
// The sphere-lattice modes: globe (searching), rubik (solving) and
// wave (listening). All draw a lat/long dot field with mode-specific
// motion, then hand off to the shared z-sorted painter.

import CoreGraphics
import Foundation

// --- the shared solver heartbeat (rubik) ------------------------------
// Rapid eased moves scramble, then replay in reverse (palindrome) so
// everything clicks back to solved, rests, repeats.

private struct Move {
    var axis: Int
    var lo: Double
    var hi: Double
    var ang: Double
}

private struct SolveCycle {
    var amount: [Double]
    var active: Int
}

private func solveCycle(_ time: Double, _ count: Int, _ slotDur: Double, _ rest: Double) -> SolveCycle {
    let cyc = 2 * Double(count) * slotDur + rest
    let tc = time.truncatingRemainder(dividingBy: cyc)
    var amount = [Double](repeating: 0, count: count)
    var active = -1
    if tc < 2 * Double(count) * slotDur {
        let slot = Int(tc / slotDur)
        let p = (tc - Double(slot) * slotDur) / slotDur
        let cl = min(1, p / 0.7)
        let ep = 1 - pow(1 - cl, 3) // machine ease-out
        if slot < count {
            for i in 0..<slot { amount[i] = 1 }
            amount[slot] = ep
            active = slot
        } else {
            let u = 2 * count - 1 - slot
            for i in 0..<u { amount[i] = 1 }
            amount[u] = 1 - ep
            active = u
        }
    }
    return SolveCycle(amount: amount, active: active)
}

private func applyMoves(
    _ p: (Double, Double, Double), _ moves: [Move], _ sc: SolveCycle
) -> (Double, Double, Double, Bool) {
    var (x, y, z) = p
    var inActive = false
    for i in 0..<moves.count {
        if sc.amount[i] <= 0 { continue }
        let mv = moves[i]
        let coord = mv.axis == 0 ? x : mv.axis == 1 ? y : z
        if coord < mv.lo || coord >= mv.hi { continue }
        if i == sc.active { inActive = true }
        let a = mv.ang * sc.amount[i]
        let ca = cos(a)
        let sa = sin(a)
        if mv.axis == 0 {
            let y2 = y * ca - z * sa
            z = y * sa + z * ca
            y = y2
        } else if mv.axis == 1 {
            let x2 = x * ca + z * sa
            z = -x * sa + z * ca
            x = x2
        } else {
            let x2 = x * ca - y * sa
            y = x * sa + y * ca
            x = x2
        }
    }
    return (x, y, z, inActive)
}

private func makeMoves(_ count: Int) -> [Move] {
    (0..<count).map { i in
        let axis = min(2, Int(hashD(Double(i), 2.3) * 3))
        let lo = -1.0 + 0.5 * Double(min(3, Int(hashD(Double(i), 5.9) * 4)))
        let dir: Double = hashD(Double(i), 7.7) < 0.5 ? 1 : -1
        return Move(axis: axis, lo: lo, hi: lo + 0.5, ang: dir * .pi / 2)
    }
}

// --- Globe: lat/long field, a scan meridian sweeps — searching --------

let drawGlobe: ModeDraw = { ctx, size, t, dark, o, tint in
    let spin = 0.5
    let cx = size / 2
    let cy = size / 2
    let radius = (size / 2) * 0.82
    let tilt = 0.4 + 0.06 * sin(t * 0.35)
    let pt = makeProj(yaw: t * spin, tilt: tilt, cx: cx, cy: cy, scale: radius)
    // scan sweeps relative to the spin; scanMul scales that relative rate
    let scan = t * (spin + (1.7 - spin) * (o["scanMul"] ?? 1))
    let rs = radiusScale(size, o["rsPow"] ?? 0.6)
    let dimBase = o["dimBase"] ?? 1

    var dots: [Dot] = []
    let latRings = Int(o["latRings"] ?? 17)
    let lonDensity = o["lonDensity"] ?? 44
    for li in 0...latRings {
        let lat = -Double.pi / 2 + (Double(li) / Double(latRings)) * .pi
        let cosLat = cos(lat)
        let sinLat = sin(lat)
        let lonCount = max(1, Int(jsRound(abs(cosLat) * lonDensity)))
        for lj in 0..<lonCount {
            let lon = (Double(lj) / Double(lonCount)) * 2 * .pi
            let (px, py, z) = pt(cosLat * cos(lon), sinLat, cosLat * sin(lon))
            let depth = (z + 1) / 2
            // the scan: a moving meridian read as a size ripple, not a shine
            let d = angleDelta(lon + t * spin, scan)
            let boost = exp(-(d * d) / 0.18) * max(0, z)
            dots.append(Dot(
                x: px, y: py, z: z,
                r: ((o["rBase"] ?? 0.6) + (o["rDepth"] ?? 1.7) * depth + (o["rBoost"] ?? 1) * boost) * rs,
                white: (o["inkFar"] ?? 0.62) - (o["inkSpan"] ?? 0.54) * depth,
                // dimBase < 1 fades un-scanned dots so the meridian reads clearly
                a: dimBase + (1 - dimBase) * min(1, boost)
            ))
        }
    }
    paint(ctx, &dots, dark: dark, rMin: o["rMin"] ?? 0.3, tint: tint)
}

// --- Rubik: bands twist in quarter turns, scramble → solve — solving --

let drawRubik: ModeDraw = { ctx, size, t, dark, o, tint in
    let cx = size / 2
    let cy = size / 2
    let R = (size / 2) * 0.82
    let pt = makeProj(yaw: t * 0.55, tilt: 0.35 + 0.1 * sin(t * 0.9), cx: cx, cy: cy, scale: R)
    let rs = radiusScale(size, o["rsPow"] ?? 0.6)
    let moveCount = Int(o["moveCount"] ?? 14)
    let moves = makeMoves(moveCount)
    let sc = solveCycle(t, moveCount, 0.42, 1.2)

    var dots: [Dot] = []
    let latRings = Int(o["latRings"] ?? 15)
    let lonDensity = o["lonDensity"] ?? 40
    for li in 0...latRings {
        let lat = -Double.pi / 2 + (Double(li) / Double(latRings)) * .pi
        let cosLat = cos(lat)
        let sinLat = sin(lat)
        let lonCount = max(1, Int(jsRound(abs(cosLat) * lonDensity)))
        for lj in 0..<lonCount {
            let lon = (Double(lj) / Double(lonCount)) * 2 * .pi
            let (x, y, z, inActive) = applyMoves((cosLat * cos(lon), sinLat, cosLat * sin(lon)), moves, sc)
            let (px, py, zr) = pt(x, y, z)
            let depth = (zr + 1) / 2
            // the band being turned inks a touch darker — the "hand"
            dots.append(Dot(
                x: px, y: py, z: zr,
                r: ((o["rBase"] ?? 0.6) + (o["rDepth"] ?? 1.7) * depth + (inActive ? (o["rActive"] ?? 0.3) : 0)) * rs,
                white: (o["inkFar"] ?? 0.62) - (o["inkSpan"] ?? 0.54) * depth - (inActive ? 0.14 : 0)
            ))
        }
    }
    paint(ctx, &dots, dark: dark, rMin: o["rMin"] ?? 0.3, tint: tint)
}

// --- Wave: a waveform rolls through the rings — listening -------------

let drawWave: ModeDraw = { ctx, size, t, dark, o, tint in
    let cx = size / 2
    let cy = size / 2
    // 0.76 base × 1.15 — the undulation pulls the sphere inward, so wave read
    // ~15% smaller than the other lattice modes; scaled up to match them
    let R = (size / 2) * 0.874
    let pt = makeProj(yaw: t * 0.18, tilt: 0.38, cx: cx, cy: cy, scale: 1)
    let rs = radiusScale(size, o["rsPow"] ?? 0.6)

    var dots: [Dot] = []
    let rings = Int(o["rings"] ?? 15)
    let lonDensity = o["lonDensity"] ?? 40
    for ri in 0...rings {
        let lat = -Double.pi / 2 + (Double(ri) / Double(rings)) * .pi
        let cosLat = cos(lat)
        let sinLat = sin(lat)
        // two waves, different tempi — organic, never quite repeating
        let w = 0.62 * sin(t * 2.1 - Double(ri) * 0.52) + 0.38 * sin(t * 1.27 + Double(ri) * 0.83)
        let rr = R * (0.88 + 0.105 * w)
        let lonCount = max(1, Int(jsRound(abs(cosLat) * lonDensity)))
        for lj in 0..<lonCount {
            let lon = (Double(lj) / Double(lonCount)) * 2 * .pi
            let (px, py, z) = pt(cosLat * cos(lon) * rr, sinLat * rr, cosLat * sin(lon) * rr)
            let depth = (z / R + 1) / 2
            let crest = max(0, w)
            dots.append(Dot(
                x: px, y: py, z: z,
                r: ((o["rBase"] ?? 0.6) + (o["rDepth"] ?? 1.7) * depth) * (1 + 0.4 * crest) * rs,
                white: 0.66 - 0.56 * depth - 0.1 * crest
            ))
        }
    }
    paint(ctx, &dots, dark: dark, rMin: o["rMin"] ?? 0.3, tint: tint)
}

File: thinking-orbs-mega/ports/svelte/src/lib/engine/lattice.ts
import {
	angleDelta,
	hashD,
	makeProj,
	paint,
	radiusScale,
	type InkColor,
} from "./core.js";
import type { Dot, ModeDraw } from "./types.js";

interface Move {
	axis: 0 | 1 | 2;
	lo: number;
	hi: number;
	ang: number;
}

function solveCycle(
	time: number,
	count: number,
	slotDuration: number,
	rest: number,
) {
	const cycle = 2 * count * slotDuration + rest;
	const cycleTime = time % cycle;
	const amount = new Array<number>(count).fill(0);
	let active = -1;
	if (cycleTime < 2 * count * slotDuration) {
		const slot = Math.floor(cycleTime / slotDuration);
		const progress = (cycleTime - slot * slotDuration) / slotDuration;
		const clamped = Math.min(1, progress / 0.7);
		const eased = 1 - (1 - clamped) ** 3;
		if (slot < count) {
			for (let i = 0; i < slot; i++) amount[i] = 1;
			amount[slot] = eased;
			active = slot;
		} else {
			const reverse = 2 * count - 1 - slot;
			for (let i = 0; i < reverse; i++) amount[i] = 1;
			amount[reverse] = 1 - eased;
			active = reverse;
		}
	}
	return { amount, active };
}

function applyMoves(
	point: [number, number, number],
	moves: Move[],
	cycle: { amount: number[]; active: number },
): [number, number, number, boolean] {
	let [x, y, z] = point;
	let inActive = false;
	for (let i = 0; i < moves.length; i++) {
		if (cycle.amount[i] <= 0) continue;
		const move = moves[i];
		const coord = move.axis === 0 ? x : move.axis === 1 ? y : z;
		if (coord < move.lo || coord >= move.hi) continue;
		if (i === cycle.active) inActive = true;
		const angle = move.ang * cycle.amount[i];
		const cosine = Math.cos(angle);
		const sine = Math.sin(angle);
		if (move.axis === 0) {
			const nextY = y * cosine - z * sine;
			z = y * sine + z * cosine;
			y = nextY;
		} else if (move.axis === 1) {
			const nextX = x * cosine + z * sine;
			z = -x * sine + z * cosine;
			x = nextX;
		} else {
			const nextX = x * cosine - y * sine;
			y = x * sine + y * cosine;
			x = nextX;
		}
	}
	return [x, y, z, inActive];
}

function makeMoves(count: number): Move[] {
	const moves: Move[] = [];
	for (let i = 0; i < count; i++) {
		const axis = Math.min(2, Math.floor(hashD(i, 2.3) * 3)) as 0 | 1 | 2;
		const lo = -1 + 0.5 * Math.min(3, Math.floor(hashD(i, 5.9) * 4));
		const direction = hashD(i, 7.7) < 0.5 ? 1 : -1;
		moves.push({ axis, lo, hi: lo + 0.5, ang: (direction * Math.PI) / 2 });
	}
	return moves;
}

export const drawGlobe: ModeDraw = (
	ctx,
	size,
	t,
	dark,
	o,
	ink: InkColor | undefined,
) => {
	const spin = 0.5;
	const center = size / 2;
	const radius = (size / 2) * 0.82;
	const tilt = 0.4 + 0.06 * Math.sin(t * 0.35);
	const project = makeProj(t * spin, tilt, center, center, radius);
	const scan = t * (spin + (1.7 - spin) * (o.scanMul ?? 1));
	const radiusMultiplier = radiusScale(size, o.rsPow ?? 0.6);
	const dimBase = o.dimBase ?? 1;
	const dots: Dot[] = [];
	const latRings = o.latRings ?? 17;
	const lonDensity = o.lonDensity ?? 44;

	for (let latIndex = 0; latIndex <= latRings; latIndex++) {
		const latitude = -Math.PI / 2 + (latIndex / latRings) * Math.PI;
		const cosineLatitude = Math.cos(latitude);
		const sineLatitude = Math.sin(latitude);
		const lonCount = Math.max(
			1,
			Math.round(Math.abs(cosineLatitude) * lonDensity),
		);
		for (let lonIndex = 0; lonIndex < lonCount; lonIndex++) {
			const longitude = (lonIndex / lonCount) * 2 * Math.PI;
			const [x, y, z] = project(
				cosineLatitude * Math.cos(longitude),
				sineLatitude,
				cosineLatitude * Math.sin(longitude),
			);
			const depth = (z + 1) / 2;
			const distance = angleDelta(longitude + t * spin, scan);
			const boost = Math.exp(-(distance * distance) / 0.18) * Math.max(0, z);
			dots.push({
				x,
				y,
				z,
				r:
					((o.rBase ?? 0.6) +
						(o.rDepth ?? 1.7) * depth +
						(o.rBoost ?? 1) * boost) *
					radiusMultiplier,
				white: (o.inkFar ?? 0.62) - (o.inkSpan ?? 0.54) * depth,
				a: dimBase + (1 - dimBase) * Math.min(1, boost),
			});
		}
	}
	paint(ctx, dots, dark, o.rMin, ink);
};

export const drawRubik: ModeDraw = (
	ctx,
	size,
	t,
	dark,
	o,
	ink: InkColor | undefined,
) => {
	const center = size / 2;
	const radius = (size / 2) * 0.82;
	const project = makeProj(
		t * 0.55,
		0.35 + 0.1 * Math.sin(t * 0.9),
		center,
		center,
		radius,
	);
	const radiusMultiplier = radiusScale(size, o.rsPow ?? 0.6);
	const moveCount = o.moveCount ?? 14;
	const moves = makeMoves(moveCount);
	const cycle = solveCycle(t, moveCount, 0.42, 1.2);
	const dots: Dot[] = [];
	const latRings = o.latRings ?? 15;
	const lonDensity = o.lonDensity ?? 40;

	for (let latIndex = 0; latIndex <= latRings; latIndex++) {
		const latitude = -Math.PI / 2 + (latIndex / latRings) * Math.PI;
		const cosineLatitude = Math.cos(latitude);
		const sineLatitude = Math.sin(latitude);
		const lonCount = Math.max(
			1,
			Math.round(Math.abs(cosineLatitude) * lonDensity),
		);
		for (let lonIndex = 0; lonIndex < lonCount; lonIndex++) {
			const longitude = (lonIndex / lonCount) * 2 * Math.PI;
			const [x, y, z, inActive] = applyMoves(
				[
					cosineLatitude * Math.cos(longitude),
					sineLatitude,
					cosineLatitude * Math.sin(longitude),
				],
				moves,
				cycle,
			);
			const [projectedX, projectedY, projectedZ] = project(x, y, z);
			const depth = (projectedZ + 1) / 2;
			dots.push({
				x: projectedX,
				y: projectedY,
				z: projectedZ,
				r:
					((o.rBase ?? 0.6) +
						(o.rDepth ?? 1.7) * depth +
						(inActive ? (o.rActive ?? 0.3) : 0)) *
					radiusMultiplier,
				white:
					(o.inkFar ?? 0.62) -
					(o.inkSpan ?? 0.54) * depth -
					(inActive ? 0.14 : 0),
			});
		}
	}
	paint(ctx, dots, dark, o.rMin, ink);
};

export const drawWave: ModeDraw = (
	ctx,
	size,
	t,
	dark,
	o,
	ink: InkColor | undefined,
) => {
	const center = size / 2;
	const radius = (size / 2) * 0.874;
	const project = makeProj(t * 0.18, 0.38, center, center, 1);
	const radiusMultiplier = radiusScale(size, o.rsPow ?? 0.6);
	const dots: Dot[] = [];
	const rings = o.rings ?? 15;
	const lonDensity = o.lonDensity ?? 40;

	for (let ring = 0; ring <= rings; ring++) {
		const latitude = -Math.PI / 2 + (ring / rings) * Math.PI;
		const cosineLatitude = Math.cos(latitude);
		const sineLatitude = Math.sin(latitude);
		const wave =
			0.62 * Math.sin(t * 2.1 - ring * 0.52) +
			0.38 * Math.sin(t * 1.27 + ring * 0.83);
		const ringRadius = radius * (0.88 + 0.105 * wave);
		const lonCount = Math.max(
			1,
			Math.round(Math.abs(cosineLatitude) * lonDensity),
		);
		for (let lonIndex = 0; lonIndex < lonCount; lonIndex++) {
			const longitude = (lonIndex / lonCount) * 2 * Math.PI;
			const [x, y, z] = project(
				cosineLatitude * Math.cos(longitude) * ringRadius,
				sineLatitude * ringRadius,
				cosineLatitude * Math.sin(longitude) * ringRadius,
			);
			const depth = (z / radius + 1) / 2;
			const crest = Math.max(0, wave);
			dots.push({
				x,
				y,
				z,
				r:
					((o.rBase ?? 0.6) + (o.rDepth ?? 1.7) * depth) *
					(1 + 0.4 * crest) *
					radiusMultiplier,
				white: 0.66 - 0.56 * depth - 0.1 * crest,
			});
		}
	}
	paint(ctx, dots, dark, o.rMin, ink);
};
